src/Stack/Setup.hs

```diff
@@ -622,46 +623,47 @@ ensureDockerStackExe containerPlatform = do
         downloadStackExe platforms sri stackExeDir (const $ return ())
     return stackExePath
 
--- | Install the newest version of Cabal globally
+-- | Install the newest version or a specific version of Cabal globally
 upgradeCabal :: (StackM env m, HasConfig env, HasGHCVariant env)
              => EnvOverride
              -> WhichCompiler
+             -> UpgradeTo
              -> m ()
-upgradeCabal menv wc = do
+upgradeCabal menv wc cabalVersion = do
     let name = $(mkPackageName "Cabal")
     rmap <- resolvePackages menv Nothing Map.empty (Set.singleton name)
-    newest <-
-        case map rpIdent rmap of
+    version <- case cabalVersion of
+        Specific v -> return v
+        Latest     -> case map rpIdent rmap of
             [] -> error "No Cabal library found in index, cannot upgrade"
-            [PackageIdentifier name' version]
-                | name == name' -> return version
+            [PackageIdentifier name' version] | name == name' -> return version
             x -> error $ "Unexpected results for resolvePackages: " ++ show x
     installed <- getCabalPkgVer menv wc
-    if installed >= newest
+    if installed >= version
         then $logInfo $ T.concat
             [ "Currently installed Cabal is "
             , T.pack $ versionString installed
-            , ", newest is "
-            , T.pack $ versionString newest
-            , ". I'm not upgrading Cabal."
+            , case cabalVersion of
+                  Specific _ -> " (specified)"
+                  Latest     -> " (latest)"
+            , ". No work to be done."
```

Comment:
This formatting corresponds to stopping the _same_ version being installed. If you want to get an old version, you just fire off `--install-cabal FOO` instead of `--upgrade-cabal`. Good question regarding the unregister/hide stuff. Will investigate.